src/nxt_cgroup.c

```diff
@@ -0,0 +1,166 @@
+/*
+ * Copyright (C) Andrew Clayton
+ * Copyright (C) F5, Inc.
+ */
+
+#include <nxt_main.h>
+
+#include <limits.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <string.h>
+#include <unistd.h>
+
+#include <nxt_cgroup.h>
+#include <nxt_errno.h>
+#include <nxt_fs.h>
+#include <nxt_malloc.h>
+#include <nxt_process.h>
+
+
+static char *nxt_mk_cgpath(const char *dir);
+
+
+nxt_int_t
+nxt_cgroup_proc_add(const nxt_task_t *task, nxt_process_t *process)
+{
+    int   ret, len;
+    FILE  *fp;
+    char  *cgpath;
+    char  cgprocs[PATH_MAX - 13];   /* - 13 for '/cgroup.procs' */
+
+    if (task->thread->runtime->type != NXT_PROCESS_MAIN
+        || nxt_process_type(process) != NXT_PROCESS_PROTOTYPE
+        || process->isolation.cgroup.path == NULL)
+    {
+        return NXT_OK;
+    }
+
+    cgpath = nxt_mk_cgpath(process->isolation.cgroup.path);
+    if (cgpath == NULL) {
+        return NXT_ERROR;
+    }
+    nxt_fs_mkdir_all((const u_char *) cgpath, 0777);
```

Comment:
You know how much I hate casts )

I just developed some macro that will make us quite happier, I think:

```c
#define char_cast(p)  _Generic(p, \
	char *: (u_char *) p, \
	const char *: (const u_char *) p, \
	u_char *: (char *) p, \
	const u_char *: (const char *) p)
```

Another reason to require GNU11.  _Generic(3) was added in C11 (although maybe compilers provide it in gnu89 as an extension, but I don't know).

Would you replace that cast (and any other casts) by that macro?

Reply:
> You know how much I hate casts )
> 
> I just developed some macro that will make us quite happier, I think:
> 
> ```c
> #define char_cast(p)  _Generic(p, \
> 	char *: (u_char *) p, \
> 	const char *: (const u_char *) p, \
> 	u_char *: (char *) p, \
> 	const u_char *: (const char *) p)
> ```

You'll need to explain to me why _all_ that is better than the simple cast. Yes, it would obviously be better to not have to cast in the first place. I would really like to know what all this u_char was about...
 
> Another reason to require GNU11. _Generic(3) was added in C11 (although maybe compilers provide it in gnu89 as an extension, but I don't know).
> 
> Would you replace that cast (and any other casts) by that macro?

Heh, I personally wouldn't, no ;) But at the very least I would probably call it cchar(p)/CCHAR(p)

In fact to fix the _few_ cases of _nxt_fs_mkdir_parent()_ & _nxt_fs_mkdir_all()_ I would go as far as

```diff
diff --git a/src/nxt_controller.c b/src/nxt_controller.c
index 99c7ca10..43371dca 100644
--- a/src/nxt_controller.c
+++ b/src/nxt_controller.c
@@ -670,7 +670,7 @@ nxt_runtime_controller_socket(nxt_task_t *task, nxt_runtime_
t *rt)
     if (ls->sockaddr->u.sockaddr.sa_family == AF_UNIX) {
         const char *path = ls->sockaddr->u.sockaddr_un.sun_path;
 
-        nxt_fs_mkdir_parent((const u_char *) path, 0755);
+        nxt_fs_mkdir_parent(path, 0755);
     }
 #endif
 
diff --git a/src/nxt_file.c b/src/nxt_file.c
index 5d38d57e..978bd053 100644
--- a/src/nxt_file.c
+++ b/src/nxt_file.c
@@ -18,7 +18,7 @@ nxt_file_open(nxt_task_t *task, nxt_file_t *file, nxt_uint_t m
ode,
     /* O_NONBLOCK is to prevent blocking on FIFOs, special devices, etc. */
     mode |= (O_NONBLOCK | create);
 
-    file->fd = open((char *) file->name, mode, access);
+    file->fd = open(file->name, mode, access);
 
     file->error = (file->fd == -1) ? nxt_errno : 0;
 
@@ -259,7 +259,7 @@ nxt_file_info(nxt_file_t *file, nxt_file_info_t *fi)
     int  n;
 
     if (file->fd == NXT_FILE_INVALID) {
-        n = stat((char *) file->name, fi);
+        n = stat(file->name, fi);
 
         file->error = (n != 0) ? nxt_errno : 0;
 
@@ -302,7 +302,7 @@ nxt_file_delete(nxt_file_name_t *name)
 {
     nxt_thread_log_debug("unlink(\"%FN\")", name);
 
-    if (nxt_fast_path(unlink((char *) name) == 0)) {
+    if (nxt_fast_path(unlink(name) == 0)) {
         return NXT_OK;
     }
 
@@ -315,7 +315,7 @@ nxt_file_delete(nxt_file_name_t *name)
 nxt_int_t
 nxt_file_set_access(nxt_file_name_t *name, nxt_file_access_t access)
 {
-    if (nxt_fast_path(chmod((char *) name, access) == 0)) {
+    if (nxt_fast_path(chmod(name, access) == 0)) {
         return NXT_OK;
     }
 
@@ -332,7 +332,7 @@ nxt_file_rename(nxt_file_name_t *old_name, nxt_file_name_t *
new_name)
 
     nxt_thread_log_debug("rename(\"%FN\", \"%FN\")", old_name, new_name);
 
-    ret = rename((char *) old_name, (char *) new_name);
+    ret = rename(old_name, new_name);
     if (nxt_fast_path(ret == 0)) {
         return NXT_OK;
     }
diff --git a/src/nxt_file.h b/src/nxt_file.h
index 07c7a22b..79e27b37 100644
--- a/src/nxt_file.h
+++ b/src/nxt_file.h
@@ -18,7 +18,7 @@ typedef struct stat                 nxt_file_info_t;
 
 #define NXT_FILE_SYSTEM_NAME_UTF8   1
 
-typedef u_char                      nxt_file_name_t;
+typedef char                        nxt_file_name_t;
 
 
 typedef struct {
diff --git a/src/nxt_fs.c b/src/nxt_fs.c
index a467da98..47546ae9 100644
--- a/src/nxt_fs.c
+++ b/src/nxt_fs.c
@@ -5,17 +5,17 @@
 #include <nxt_main.h>
 
 
-static nxt_int_t nxt_fs_mkdir(const u_char *dir, mode_t mode);
+static nxt_int_t nxt_fs_mkdir(const char *dir, mode_t mode);
 
 
 nxt_int_t
-nxt_fs_mkdir_all(const u_char *dir, mode_t mode)
+nxt_fs_mkdir_all(const char *dir, mode_t mode)
 {
     char    *start, *end, *dst;
     size_t  dirlen;
     char    path[PATH_MAX];
 
-    dirlen = nxt_strlen(dir);
+    dirlen = strlen(dir);
 
     nxt_assert(dirlen < PATH_MAX && dirlen > 1 && dir[0] == '/');
 
@@ -35,7 +35,7 @@ nxt_fs_mkdir_all(const u_char *dir, mode_t mode)
         dst = nxt_cpymem(dst, start, end - start);
         *dst = '\0';
 
-        if (nxt_slow_path(nxt_fs_mkdir((u_char *) path, mode) != NXT_OK
+        if (nxt_slow_path(nxt_fs_mkdir(path, mode) != NXT_OK
                           && nxt_errno != EEXIST))
         {
             return NXT_ERROR;
@@ -49,12 +49,12 @@ nxt_fs_mkdir_all(const u_char *dir, mode_t mode)
 
 
 nxt_int_t
-nxt_fs_mkdir_parent(const u_char *path, mode_t mode)
+nxt_fs_mkdir_parent(const char *path, mode_t mode)
 {
     char       *ptr, *dir;
     nxt_int_t  ret;
 
-    dir = nxt_strdup(path);
+    dir = strdup(path);
     if (nxt_slow_path(dir == NULL)) {
         return NXT_ERROR;
     }
@@ -64,7 +64,7 @@ nxt_fs_mkdir_parent(const u_char *path, mode_t mode)
     ptr = strrchr(dir, '/');
     if (nxt_fast_path(ptr != NULL)) {
         *ptr = '\0';
-        ret = nxt_fs_mkdir((const u_char *) dir, mode);
+        ret = nxt_fs_mkdir(dir, mode);
     }
 
     nxt_free(dir);
@@ -74,9 +74,9 @@ nxt_fs_mkdir_parent(const u_char *path, mode_t mode)
 
 
 static nxt_int_t
-nxt_fs_mkdir(const u_char *dir, mode_t mode)
+nxt_fs_mkdir(const char *dir, mode_t mode)
 {
-    if (nxt_fast_path(mkdir((const char *) dir, mode) == 0)) {
+    if (nxt_fast_path(mkdir(dir, mode) == 0)) {
         return NXT_OK;
     }
 
diff --git a/src/nxt_fs.h b/src/nxt_fs.h
index c8868d80..054dd00e 100644
--- a/src/nxt_fs.h
+++ b/src/nxt_fs.h
@@ -6,8 +6,8 @@
 #define _NXT_FS_H_INCLUDED_
 
 
-nxt_int_t nxt_fs_mkdir_parent(const u_char *path, mode_t mode);
-nxt_int_t nxt_fs_mkdir_all(const u_char *dir, mode_t mode);
+nxt_int_t nxt_fs_mkdir_parent(const char *path, mode_t mode);
+nxt_int_t nxt_fs_mkdir_all(const char *dir, mode_t mode);
 
 
 #endif  /* _NXT_FS_H_INCLUDED_ */
diff --git a/src/nxt_http_static.c b/src/nxt_http_static.c
index 0507e038..000e53e5 100644
--- a/src/nxt_http_static.c
+++ b/src/nxt_http_static.c
@@ -366,7 +366,7 @@ nxt_http_static_send_ready(nxt_task_t *task, void *obj, void *data)
 
     nxt_memzero(&file, sizeof(nxt_file_t));
 
-    file.name = fname;
+    file.name = (char *) fname;
 
 #if (NXT_HAVE_OPENAT2)
     if (conf->resolve != 0 || ctx->chroot.length > 0) {
@@ -384,10 +384,11 @@ nxt_http_static_send_ready(nxt_task_t *task, void *obj, void *data)
 
             fname = share->is_const
                     ? share->fname
-                    : nxt_http_static_chroot_match(chr->start, file.name);
+                    : nxt_http_static_chroot_match(chr->start,
+                                                   (u_char *) file.name);
 
             if (fname != NULL) {
-                file.name = chr->start;
+                file.name = (char *) chr->start;
                 ret = nxt_file_open(task, &file, NXT_FILE_SEARCH, NXT_FILE_OPEN,
                                     0);
 
@@ -397,11 +398,11 @@ nxt_http_static_send_ready(nxt_task_t *task, void *obj, void *data)
             }
 
         } else if (fname[0] == '/') {
-            file.name = (u_char *) "/";
+            file.name = (char *) "/";
             ret = nxt_file_open(task, &file, NXT_FILE_SEARCH, NXT_FILE_OPEN, 0);
 
         } else {
-            file.name = (u_char *) ".";
+            file.name = (char *) ".";
             file.fd = AT_FDCWD;
             ret = NXT_OK;
         }
@@ -411,7 +412,7 @@ nxt_http_static_send_ready(nxt_task_t *task, void *obj, void *data)
 
             af = file;
             nxt_memzero(&file, sizeof(nxt_file_t));
-            file.name = fname;
+            file.name = (char *) fname;
 
             ret = nxt_file_openat2(task, &file, NXT_FILE_RDONLY,
                                    NXT_FILE_OPEN, 0, af.fd, resolve);
diff --git a/src/nxt_isolation.c b/src/nxt_isolation.c
index 796da4c6..c86f99ac 100644
--- a/src/nxt_isolation.c
+++ b/src/nxt_isolation.c
@@ -726,7 +726,7 @@ nxt_isolation_prepare_rootfs(nxt_task_t *task, nxt_process_t *process)
             continue;
         }
 
-        ret = nxt_fs_mkdir_all(dst, S_IRWXU | S_IRWXG | S_IRWXO);
+        ret = nxt_fs_mkdir_all((const char *) dst, S_IRWXU | S_IRWXG | S_IRWXO);
         if (nxt_slow_path(ret != NXT_OK)) {
             nxt_alert(task, "mkdir(%s) %E", dst, nxt_errno);
             goto undo;

diff --git a/src/nxt_runtime.c b/src/nxt_runtime.c
index c7e4455e..7ce0d719 100644
--- a/src/nxt_runtime.c
+++ b/src/nxt_runtime.c
@@ -882,7 +882,7 @@ nxt_runtime_conf_init(nxt_task_t *task, nxt_runtime_t *rt)
 
     if (nxt_fast_path(ret == 0 || nxt_errno == EEXIST)) {
         rt->certs.length = file_name.len;
-        rt->certs.start = file_name.start;
+        rt->certs.start = (u_char *) file_name.start;
 
     } else {
         nxt_alert(task, "Unable to create certificates storage directory: "
diff --git a/src/nxt_sprintf.c b/src/nxt_sprintf.c
index 9c8e27ed..d925dd36 100644
--- a/src/nxt_sprintf.c
+++ b/src/nxt_sprintf.c
@@ -109,6 +109,7 @@ nxt_vsprintf(u_char *buf, u_char *end, const char *fmt, va_list args)
     nxt_nsec_t           ns;
     nxt_bool_t           sign;
     const u_char         *p;
+    const char           *s;
     nxt_sprintf_t        spf;
     nxt_file_name_t      *fn;
 
@@ -474,10 +475,10 @@ nxt_vsprintf(u_char *buf, u_char *end, const char *fmt, va_list args)
 
             case 'N':
                 fn = va_arg(args, nxt_file_name_t *);
-                p = fn;
+                s = fn;
 
-                while (*p != '\0' && buf < end) {
-                    *buf++ = *p++;
+                while (*s != '\0' && buf < end) {
+                    *buf++ = *s++;
                 }
 
                 fmt++;
```

I had to add a few casts and change a few. At some point you hit up against our old friend _nxt\_str\_t_

Reply:
> > You know how much I hate casts )
> > I just developed some macro that will make us quite happier, I think:
> > ```c
> > #define char_cast(p)  _Generic(p, \
> > 	char *: (u_char *) p, \
> > 	const char *: (const u_char *) p, \
> > 	u_char *: (char *) p, \
> > 	const u_char *: (const char *) p)
> > ```
> 
> You'll need to explain to me why _all_ that is better than the simple cast.

Because the macro guarantees that you don't cast from `struct iovec *` to `char *`.  More or less, the rationale of C++'s `const_cast()` (which is the inspiration of this macro).  I need to say here that I don't defend C++'s casts; it's a consequence of a broken design in the language, which started by removing the implicit conversions of `void *`, and ended in the braindamage that it is right now.

> Yes, it would obviously be better to not have to cast in the first place. I would really like to know what all this u_char was about...

Yep, of course.

BTW, after some real-world testing, I further developed this:

```c
#define nxt_char_cast(p)                                                      \
    _Generic(&*p,                                                             \
    u_char *:        (char *) p,                                              \
    u_char **:       (char **) p,                                             \
    const u_char *:  (const char *) p,                                        \
    default:         p)


#define nxt_uchar_cast(p)                                                     \
    _Generic(&*p,                                                             \
    char *:        (u_char *) p,                                              \
    char **:       (u_char **) p,                                             \
    const char *:  (const u_char *) p,                                        \
    default:       p)
```

> 
> > Another reason to require GNU11. _Generic(3) was added in C11 (although maybe compilers provide it in gnu89 as an extension, but I don't know).
> > Would you replace that cast (and any other casts) by that macro?
> 
> Heh, I personally wouldn't, no ;) But at the very least I would probably call it cchar(p)/CCHAR(p)

I used `char_cast()` because it was inspired by the following [`const_cast()`](https://software.codidact.com/posts/286575/287345#answer-287345) that I wrote, which itself was inspired by an `optional_cast()` to remove a hypothetical `_Optional` qualifier (a pre-proposal for ISO C about nonnull pointers), which itself was inspired by C++'s `const_cast()`.

The idea is to be consistent with `xxx_cast()` well-understood syntax.

> 
> In fact to fix the _few_ cases of _nxt_fs_mkdir_parent()_ & _nxt_fs_mkdir_all()_ I would go as far as
> 
> ```diff
> diff --git a/src/nxt_controller.c b/src/nxt_controller.c
> index 99c7ca10..43371dca 100644
> --- a/src/nxt_controller.c
> +++ b/src/nxt_controller.c
> @@ -670,7 +670,7 @@ nxt_runtime_controller_socket(nxt_task_t *task, nxt_runtime_
> t *rt)
>      if (ls->sockaddr->u.sockaddr.sa_family == AF_UNIX) {
>          const char *path = ls->sockaddr->u.sockaddr_un.sun_path;
>  
> -        nxt_fs_mkdir_parent((const u_char *) path, 0755);
> +        nxt_fs_mkdir_parent(path, 0755);
>      }
>  #endif
>  ```

[...]

> 
> I had to add a few casts and change a few. At some point you hit up against our old friend _nxt_str_t_

Yeah, I don't like that change, until _nxt_str_t_ is also changed (not any soon, we know).

Reply:
> BTW, after some real-world testing, I further developed this:
> 
> ```c
> #define nxt_char_cast(p)                                                      \
>     _Generic(&*p,                                                             \
>     u_char *:        (char *) p,                                              \
>     u_char **:       (char **) p,                                             \
>     const u_char *:  (const char *) p,                                        \
>     default:         p)
> 
> 
> #define nxt_uchar_cast(p)                                                     \
>     _Generic(&*p,                                                             \
>     char *:        (u_char *) p,                                              \
>     char **:       (u_char **) p,                                             \
>     const char *:  (const u_char *) p,                                        \
>     default:       p)
> ```

That at least looks a little more palatable ;)

Reply:
I have a third round, which is more readable, albeit more verbose in call sites.  It is also simpler to use, since there's only a single macro again (as in the first iteration).  This is what I first tried to implement, but I didn't come up with some of the small details:

```c
#if (NXT_HAVE_C11_GENERIC)
#define nxt_char_cast(t, p)                                                   \
    _Generic(&*(p),                                                           \
    u_char *:                                                                 \
        _Generic((t)NULL, char *: (t)(p), const char *: (t)(p), default: (p)),\
    u_char **:                                                                \
        _Generic((t)NULL, char **: (t)(p), default: (p)),                     \
    const u_char *:                                                           \
        _Generic((t)NULL, const char *: (t)(p), default: (p)),                \
    char *:                                                                   \
        _Generic((t)NULL, u_char *: (t)(p), const u_char *: (t)(p), default: (p)),\
    char **:                                                                  \
        _Generic((t)NULL, u_char **: (t)(p), default: (p)),                   \
    const char *:                                                             \
        _Generic((t)NULL, const u_char *: (t)(p), default: (p)),              \
    default:                                                                  \
        (p)                                                                   \
    )
#else
#define nxt_char_cast(t, p)  ((t) (p))
#endif
```

To be used like

```c
#define nxt_strlen(s)  strlen(nxt_char_cast(const char *, (s)))
```

This macro deserves some explanation, since it's by far the most complex use of _Generic(3) I've seen so far.  In fact, this example will go to the recently-added _Generic(3) manual page:
<https://manpages.debian.org/testing/manpages-dev/_Generic.3.en.html>

- It requires the caller to tell which type is wanted after the cast.  This is a bit verbose, but it helps readability and greppability, since it allows finding all casts to say `char **` trivially, and you can also tell to which type it is converting just by reading the call site, so you don't need to check which type the pointer being casted originally had.

The main concepts behind this cast are:
-  Allow adding `const` to non-`const` pointers (but not pointers to pointers).
-  Allow no-op casts, that is, converting `char *` to `char *`.

If _Generic(3) is not available, just do the cast blindly.

---

This cast is not only interesting in our code base, due to the dubious use of `u_char`.  It is actually a very interesting cast for general programming, when needing to use syscalls like bind(2).  Fundamentally, this cast macro is a rudimentary version of C++'s `static_cast()`.  You could develop a similar (and that's the case I'll be adding to the manual page) sockaddr_cast() to call bind(2) safely.